extensions/oplogPopulator/OplogPopulator.js

```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+     * Deletes unused connectors that were
+     * previously created by this oplogPopulator
+     * @returns {undefined}
+     */
+    async _deleteUnusedConnectors() {
+        // getting all active connectors
+        const activeConnectors = await this._connectWrapper.getConnectors();
+        // filtering out used connectors & connectors that don't
+        // respect the naming of connectors
+        const unusedConnectors = activeConnectors.filter(connector => {
+            const used = this._configuredConnectors.includes(connector);
+            const namePrefix = this._config.prefix || '';
+            const startsWithPrefix = connector.startsWith(`${namePrefix}${constants.defaultConnectorName}-`);
+            return [!used, startsWithPrefix].every(x => x);
+        });
+        // deleting unused connectors
+        await Promise.all(unusedConnectors.map(connector => this._connectWrapper.deleteConnector(connector)));
+        this._logger.debug('Deleted unused buckets', {
+            method: 'OplogPopulator._deleteUnusedConnectors',
+            connectors: unusedConnectors
+        });
+    }
+
+    /**
+     * Sets the OplogPopulator
      * @returns {Promise|undefined} undefined
      * @throws {InternalError}
      */
     async setup() {
        try {
-           await this._setupMongoClient();
-           // getting buckets with at least one extension active
-           const backbeatEnabledBuckets = await this._getBackbeatEnabledBuckets();
-           this._bucketsForConnector = new Set(backbeatEnabledBuckets);
-           // TODO: add support for multiple connectors
-            const defaultConnectorConfig = this._getDefaultConnectorConfiguration(constants.defaultConnectorName);
-            defaultConnectorConfig.pipeline = this._generateNewConnectorPipeline(backbeatEnabledBuckets);
-            const connectorConfigs = {
-                name: constants.defaultConnectorName,
-                config: defaultConnectorConfig
-            };
-            await this._configureConnector(connectorConfigs);
-           this._setMetastoreChangeStream();
+            // initializing the mongo client
+            await this._setupMongoClient();
+            // generating & saving connector names
+            this._setConnectorNames();
+            // initializing data store
+            this._configuredConnectors.forEach(connector => {
+                this._bucketsForConnectors[connector] = new Set();
+            });
+            // getting buckets with at least one backbeat extension active
+            const backbeatEnabledBuckets = await this._getBackbeatEnabledBuckets();
+            // saving previous state of connectors
+            await this._restoreConnectorBuckets(backbeatEnabledBuckets);
+            // assigning a connector to bucket not in previous state
+            this._handleBucketsWithNoConnectors(backbeatEnabledBuckets);
+            // creating/updating the connectors
+            await this._spawnConnectors();
```

Comment:
may be created on demand, which would provide more future-proof interface with the "allocation strategy" (i.e. if/when we want to support a strategy with max-number-of-buckets-per-connector)